src/Neo/SmartContract/Native/NeoToken.cs

```diff
@@ -327,13 +327,34 @@ public BigInteger UnclaimedGas(DataCache snapshot, UInt160 account, uint end)
             return CalculateBonus(snapshot, state, end);
         }
 
+        [ContractMethod(Hardfork.HF_Echidna, RequiredCallFlags = CallFlags.States | CallFlags.AllowCall | CallFlags.AllowNotify)]
+        private async ContractTask OnNEP17Payment(ApplicationEngine engine, UInt160 from, BigInteger amount, StackItem data)
+        {
+            if (engine.CallingScriptHash != GAS.Hash)
+                throw new InvalidOperationException("only GAS is accepted");
+
+            if ((long)amount != GetRegisterPrice(engine.SnapshotCache))
+                throw new ArgumentException("incorrect GAS amount for registration");
+
+            var pubkey = ECPoint.DecodePoint(data.GetSpan(), ECCurve.Secp256r1);
+
+            await GAS.Burn(engine, Hash, amount);
+
+            RegisterInternal(engine, pubkey);
+        }
+
         [ContractMethod(RequiredCallFlags = CallFlags.States)]
         private bool RegisterCandidate(ApplicationEngine engine, ECPoint pubkey)
         {
-            if (!engine.CheckWitnessInternal(Contract.CreateSignatureRedeemScript(pubkey).ToScriptHash()))
-                return false;
             // In the unit of datoshi, 1 datoshi = 1e-8 GAS
             engine.AddFee(GetRegisterPrice(engine.SnapshotCache));
```

Comment:
why not return false if `AddFee` is not possible?

Reply:
How can it help? We need GAS. It either comes from the system fee or via a transfer. The first one burns the fee irrespective of the outcome. The second is what we have here.

The only other option is `GAS.Transfer(pub_key_owner, ...)` or even direct `GAS.Burn(pub_key_owner, ...)` call from NEO contract itself, but that's not the NEP-27 way. Transfers should be transfers and NEO deducting something from our GAS balance doesn't look nice.

Reply:
Por testing you can try and for execute you can not try 😆

Reply:
I mean if the problem is the gas burn, we can find a global solution for it, maybe we have this problem in a different contract

Reply:
We can add extra gas for invokes that can only be burned, not consumed by opcodes. If the execution is ApplicationTest or similar allow this behavior, and if this value is negative, return an error by rpc

Reply:
The behaviour is different, before if was wrong signed, the fee was not burned